# Day 4: Colab 4D - From Field Sheet to DataFrame

Partners: your names here

Date: 09/03/2026

In [1]:
import pandas as pd
import numpy as np

In [ ]:
url = 'https://eds-217-essential-python.github.io/data/messy_field_survey.csv'
data = pd.read_csv(url)

In [3]:
data.head()

,site,collection date,temperature_c,pH,dissolved_oxygen_mg_L,conductivity_uS_cm,n_replicates
0,SITE_F,2025-07-28,24.1,5.89,5.51,899.1,4.0
1,site-d,2025-08-05,12.5,7.74,9.97,300.1,4.0
2,Site_C,2025-07-04,22.6,6.42,7.19,815.8,4.0
3,Site_D,2025-08-15,16.0,7.74,9.73,250.7,3.0
4,site_a,2025-06-24,14.7,"7,64",9.19,338.8,3.0


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 320 entries, 0 to 319
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   site                   320 non-null    object 
 1   collection date        320 non-null    object 
 2   temperature_c          302 non-null    float64
 3   pH                     320 non-null    object 
 4   dissolved_oxygen_mg_L  309 non-null    float64
 5   conductivity_uS_cm     313 non-null    float64
 6   n_replicates           315 non-null    float64
dtypes: float64(4), object(3)
memory usage: 17.6+ KB


"collection_date" and "pH" is an object when we expect it would be a int. 
"n_replicates" is a float when we expect it would be an int.

In [6]:
data.isnull().sum()

site                      0
collection date           0
temperature_c            18
pH                        0
dissolved_oxygen_mg_L    11
conductivity_uS_cm        7
n_replicates              5
dtype: int64

temperature_c, dissolved_oxygen, conductivity, and n_replicates have gaps-- 18, 11, 7, and 5 gaps respectively.

In [ ]:
data.duplicated().sum()
#20

20

In [10]:
data['site'].unique()

array(['SITE_F', 'site-d', 'Site_C', 'Site_D', ' site_a', 'SITE_A', 'site_a',
       'SITE_D', ' site_d', 'site_c', 'site_a ', 'site-a', 'SITE_C', 'site_d',
       'site-e', 'Site_E', 'SITE_B', 'site_e ', ' Site_B', 'Site_F', 'site_f ',
       'site_f', 'site_e', 'site_c ', 'site-b', 'site_b ', 'SITE_E', 'site-f',
       'site_d ', ' site-c', 'Site_A', 'site-c', ' site-f', 'Site_B',
       ' SITE_E', 'site_b'], dtype=object)

In [11]:
data['temperature_c'].describe()

count    302.000000
mean     -11.206623
std      173.437298
min     -999.000000
25%       16.700000
50%       18.600000
75%       21.400000
max       26.500000
Name: temperature_c, dtype: float64

### Clean it !


In [12]:
data.shape

(320, 7)

In [20]:
data_dedup = data.drop_duplicates() 
data_dedup
## 300 rows|

,site,collection date,temperature_c,pH,dissolved_oxygen_mg_L,conductivity_uS_cm,n_replicates
0,SITE_F,2025-07-28,24.1,5.89,5.51,899.1,4.0
1,site-d,2025-08-05,12.5,7.74,9.97,300.1,4.0
2,Site_C,2025-07-04,22.6,6.42,7.19,815.8,4.0
3,Site_D,2025-08-15,16.0,7.74,9.73,250.7,3.0
4,site_a,2025-06-24,14.7,"7,64",9.19,338.8,3.0
...,...,...,...,...,...,...,...
315,SITE_B,2025-06-12,17.3,7.15,8.46,452.3,3.0
316,SITE_E,2025-07-24,21.0,6.58,7.08,607.5,4.0
317,site_d,2025-07-24,NaN,7.83,9.85,317.5,2.0
318,site_c,2025-08-25,20.8,6.7,7.27,767.7,3.0


In [30]:
data_dedup['site'] = data_dedup['site'].str.lower()  
data_dedup['site'] = data_dedup['site'].str.strip() 
data_dedup['site'] = data_dedup['site'].str.replace('site-', 'site_')
data_dedup['site'].value_counts()

<positron-console-cell-30>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
<positron-console-cell-30>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
<positron-console-cell-30>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


site
site_f    50
site_d    50
site_c    50
site_a    50
site_e    50
site_b    50
Name: count, dtype: int64

In [37]:
data_dedup['pH'] = data_dedup['pH'].str.replace(',', '.').astype(float)
data_dedup.describe()

<positron-console-cell-37>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


,temperature_c,pH,dissolved_oxygen_mg_L,conductivity_uS_cm,n_replicates
count,282.000000,300.000000,289.000000,293.000000,295.000000
mean,-13.368440,7.018533,8.116886,555.378498,2.969492
std,179.304155,0.492886,1.562091,232.670408,0.619209
min,-999.000000,5.780000,4.140000,186.800000,2.000000
25%,16.700000,6.615000,6.990000,320.400000,3.000000
50%,18.600000,7.020000,8.110000,540.200000,3.000000
75%,21.400000,7.372500,9.350000,781.800000,3.000000
max,26.500000,8.330000,11.430000,981.700000,4.000000


In [ ]:
cols = ['temperature_c', 'dissolved_oxygen_mg_L', 'conductivity_uS_cm']
data_dedup = data_dedup.dropna(subset=cols)
data_dedup.shape
#36

(264, 7)

In [59]:
data_dedup['n_replicates'] = data_dedup['n_replicates'].fillna(1).astype(int)
data_dedup['n_replicates'].value_counts(ascending=False)

n_replicates
3    158
2     55
4     46
1      5
Name: count, dtype: int64

In [63]:
# df_filtered = df[df['column'] > x]
data_filtered = data_dedup[data_dedup['temperature_c'] > -100].copy()
data_filtered
#9

,site,collection date,temperature_c,pH,dissolved_oxygen_mg_L,conductivity_uS_cm,n_replicates
0,site_f,2025-07-28,24.1,5.89,5.51,899.1,4
1,site_d,2025-08-05,12.5,7.74,9.97,300.1,4
2,site_c,2025-07-04,22.6,6.42,7.19,815.8,4
3,site_d,2025-08-15,16.0,7.74,9.73,250.7,3
4,site_a,2025-06-24,14.7,7.64,9.19,338.8,3
...,...,...,...,...,...,...,...
313,site_c,2025-06-20,23.3,6.54,6.43,866.6,3
315,site_b,2025-06-12,17.3,7.15,8.46,452.3,3
316,site_e,2025-07-24,21.0,6.58,7.08,607.5,4
318,site_c,2025-08-25,20.8,6.70,7.27,767.7,3


In [69]:
data_filtered['temperature_c'].describe()

count    255.000000
mean      19.218824
std        3.184392
min       12.300000
25%       16.900000
50%       19.100000
75%       21.600000
max       26.500000
Name: temperature_c, dtype: float64

The new mean is 19.22, whereas the old mean was -11.21.

In [75]:
data_rename = data_filtered.rename(columns={'collection date': 'collection_date'})
data_rename.head(1)

,site,collection_date,temperature_c,pH,dissolved_oxygen_mg_L,conductivity_uS_cm,n_replicates
0,site_f,2025-07-28,24.1,5.89,5.51,899.1,4


In [82]:
data_rename['conductivity_mS_cm'] = data_rename['conductivity_uS_cm'] / 1000

In [89]:
data_rename['temperature_f'] = (data_rename['temperature_c'] * 9 / 5) + 32
data_rename.head()

,site,collection_date,temperature_c,pH,dissolved_oxygen_mg_L,conductivity_uS_cm,n_replicates,conductivity_mS_cm,temperature_f
0,site_f,2025-07-28,24.1,5.89,5.51,899.1,4,0.8991,75.38
1,site_d,2025-08-05,12.5,7.74,9.97,300.1,4,0.3001,54.50
2,site_c,2025-07-04,22.6,6.42,7.19,815.8,4,0.8158,72.68
3,site_d,2025-08-15,16.0,7.74,9.73,250.7,3,0.2507,60.80
4,site_a,2025-06-24,14.7,7.64,9.19,338.8,3,0.3388,58.46


In [90]:
def celsius_to_fahrenheit(celsius):
    fahrenheit = (celsius * 9 / 5) + 32
    return fahrenheit

data_rename['temperature_f'] = data_rename['temperature_c'].apply(celsius_to_fahrenheit)
data_rename.head()

,site,collection_date,temperature_c,pH,dissolved_oxygen_mg_L,conductivity_uS_cm,n_replicates,conductivity_mS_cm,temperature_f
0,site_f,2025-07-28,24.1,5.89,5.51,899.1,4,0.8991,75.38
1,site_d,2025-08-05,12.5,7.74,9.97,300.1,4,0.3001,54.50
2,site_c,2025-07-04,22.6,6.42,7.19,815.8,4,0.8158,72.68
3,site_d,2025-08-15,16.0,7.74,9.73,250.7,3,0.2507,60.80
4,site_a,2025-06-24,14.7,7.64,9.19,338.8,3,0.3388,58.46


In [95]:
def classify_ph(pH):
    if pH < 6.5:
        return 'acidic'
    elif pH > 7.5:
        return 'alkaline'
    else:
        return 'neutral'

data_rename['ph_class'] = data_rename['pH'].apply(classify_ph)
data_rename['ph_class'].value_counts()

ph_class
neutral     171
acidic       42
alkaline     42
Name: count, dtype: int64

In [99]:
data_acidic = data_rename[data_rename['ph_class'] == "acidic"].copy()
data_acidic['site'].value_counts()

site
site_f    27
site_c    14
site_e     1
Name: count, dtype: int64

In [107]:
site_d = data_rename[data_rename['site'] == "site_d"]
site_f = data_rename[data_rename['site'] == "site_f"]
print("Site D Mean= ", site_d['dissolved_oxygen_mg_L'].mean())
print("Site F Mean= ", site_f['dissolved_oxygen_mg_L'].mean(), "#low")

Site D Mean=  10.053333333333333
Site F Mean=  6.180232558139535 #low


### To the PI who oversaw the collection of this data: fix the site entry standards, plz.